airflow/sensors/base.py

```diff
@@ -256,14 +256,13 @@ def _get_next_poke_interval(
         return new_interval
 
     def prepare_for_execution(self) -> BaseOperator:
-        task = super().prepare_for_execution()
         # Sensors in `poke` mode can block execution of DAGs when running
         # with single process executor, thus we change the mode to`reschedule`
         # to allow parallel task being scheduled and executed
         if conf.get("core", "executor") == "DebugExecutor":
             self.log.warning("DebugExecutor changes sensor mode to 'reschedule'.")
-            task.mode = "reschedule"
-        return task
+            self.mode = "reschedule"
```

Comment:
doing the change before calling the `super()` method, which returns a copy of `self`, so that the change is permanent.
I don't really understand why we return a copy of the task here, so there might be implications to this change that I'm missing.

Reply:
The task is copied so things applied to it later (template rendering, task-level hooks) are not carried over when the task is run next time.

Why does this `mode` change need to be permanent?

Reply:
In `TaskInstance`, there’s a line that assigns the copied task to `task`:

https://github.com/apache/airflow/blob/4e545c6e54712eedb6ca9cbb8333393ae3f6cba2/airflow/models/taskinstance.py#L1373

So why is `ti.task` here still referencing the original task?

---

Regardless of the problem above, I think it’s a bad idea to modify the original task in any case. Maybe a better alternative would be to simply add the DebugExecutor exception in ReadyToRescheduleDep so we don’t need to change the value assigned to `reschedule` at all.

Reply:
Ok, I'll try to dig to understand why `ti.task` is the original task, and look at checking the executor there if that doesn't lead me anywhere.

Reply:
so, that's because the TaskInstance that goes through `ready_to_reschedule` is not the same as the one that does the prepare_for_execution, it's a clone that's reloaded from SQL, with the original unmodified task.

Reply:
ok, new commit, I check the executor type in `ready_to_reschedule` instead, it seems to do the job well.